mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -436,39 +438,48 @@ fn find_pointers_and_order_post_crash(file: &mut File, max_file_size: u64) -> Ri
             read = end + data_size;
         }
 
-        // Found the last write, take whatever we got for read and write
-        // and return the pointers.
+        // Check next block
+        write_updates += 1;
+        order += 1; // We want to start with the next order number.
+        start = end + data_size;
+        end = start + block_size;
+
+        // Found the last write, can stop updating write.
         if inner.order() < order {
+            should_update_write = false;
+            // Update write one last time to point at block
+            // where order is less.
+            write = start % max_file_size;
+        }
+        if should_update_write {
+            write = start % max_file_size;
+        }
+
+        if start >= max_file_size {
             return RingBufferMetadata {
                 file_pointers: FilePointers {
-                    write: start,
+                    write,
                     read_begin: read,
                     read_end: read,
                 },
                 order,
                 // Check to see if wrap around case.
-                can_read_from_wrap_around: (start == read && write_updates > read_updates),
+                can_read_from_wrap_around: (write == read && write_updates > read_updates),
```

Comment:
Maybe the variable is poorly named... the particular case is for when write wraps around is at the 'exact' position that read is at. Because I don't allow for read or write pointers to pass one another, this is for allowing a special case where it is actually okay because we wrote a full buffer and can read.